cpp/src/arrow/builder-benchmark.cc

```diff
@@ -61,4 +61,70 @@ static void BM_BuildVectorNoNulls(
 
 BENCHMARK(BM_BuildVectorNoNulls)->Repetitions(3)->Unit(benchmark::kMillisecond);
 
+static void BM_BuildAdaptiveIntNoNulls(
+    benchmark::State& state) {  // NOLINT non-const reference
+  int64_t size = static_cast<int64_t>(std::numeric_limits<int16_t>::max()) * 256;
+  int64_t chunk_size = size / 8;
+  std::vector<int64_t> data;
+  for (int64_t i = 0; i < size; i++) {
+    data.push_back(i);
+  }
+  while (state.KeepRunning()) {
+    AdaptiveIntBuilder builder(default_memory_pool());
+    for (int64_t i = 0; i < size; i += chunk_size) {
+      // Build up an array of 512 MiB in size
+      builder.Append(data.data() + i, chunk_size, nullptr);
+    }
+    std::shared_ptr<Array> out;
+    builder.Finish(&out);
+  }
+  state.SetBytesProcessed(state.iterations() * data.size() * sizeof(int64_t));
+}
+
+BENCHMARK(BM_BuildAdaptiveIntNoNulls)->Repetitions(3)->Unit(benchmark::kMillisecond);
+
+static void BM_BuildAdaptiveIntNoNullsScalarAppend(
+    benchmark::State& state) {  // NOLINT non-const reference
+  int64_t size = static_cast<int64_t>(std::numeric_limits<int16_t>::max()) * 256;
+  std::vector<int64_t> data;
+  for (int64_t i = 0; i < size; i++) {
+    data.push_back(i);
+  }
+  while (state.KeepRunning()) {
+    AdaptiveIntBuilder builder(default_memory_pool());
+    for (int64_t i = 0; i < size; i++) {
+      builder.Append(data[i]);
```

Comment:
To make this benchmark more "realistic" we should add an assertion about the result of `Append` being OK